### DISCUSSION TITLE
Use of let-else syntax.

### DISCUSSION THREAD

**Discussion by Ben-PH** [massalabs/massa#3494] [Code Style]
as of rust 1.65, [let-else](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements) syntax has been stabalised. At times, this can make the code more readable:
conisider this function: https://github.com/massalabs/massa/blob/61f7bb9b3fba63a7c78b4181cda276c599fee45a/massa-client/src/cmds.rs#L1435 
```rust
    if let Some(value) = p.get_key_value(key) {
        value
            .1
            .parse::<T>()
            .map(Option::Some)
            .map_err(|e| anyhow!("failed to parse \"{}\" due to: {}", value.1, e))
    } else {
        Ok(None)
    }
```

becomes 

```rust
    let Some(value) = p.get_key_value(key) else {
        return Ok(None);
    };

    value
        .1
        .parse::<T>()
        .map(Option::Some)
        .map_err(|e| anyhow!("failed to parse \"{}\" due to: {}", value.1, e))
```

A PR clearing the `clippy::manual-let-else` lints which make these recommendations can be found in #3499

**Comment by AurelienFT** [comment]
I find it so strange to don't have the two cases of a Result in two separate lines each case making his own work. I'm so used to read let [variable_name] that this look strange to me but it's interesting in case of complex code in one branch it avoid writing everything in the branch.

I'm not fully convicted but I see some advantages

**Reply by Ben-PH** [reply]
Thanks. I found it strange at first as well, and I don't think it's in any way controversial to say that it can be very hard to justify as "better" in some cases. One example that comes to mind:

```rust
let Some(foo_field) = match foo_wrapper {
    Some(foo) => Some(foo.some_field),
    None => return;
};

// changes to:
let Some(foo_field) = foo_wrapper.map(|f| f.some_field) else {
    return;
};
```

Such a case is probably unreasonable to recommend, let alone "enforce" the use of let-else syntax.

One example that I think is particularly good use for let-else: https://github.com/massalabs/massa/pull/3499/files#diff-999f2be25e86f9ed087021cc5ae7d3e84af75602f84bfa60f6284590c2283958

I think it would be valuable to make sure we are all aware of it, and leave it up to PR authors/reviews to make a judgement call on when/where to use it.

**Reply by AurelienFT** [reply]
Yes. I tag @massalabs/core-team

**Reply by Eitu33** [reply]
I believe the best behaviour depends on the case
